test/prefer-set-has.js

```diff
@@ -0,0 +1,208 @@
+import test from 'ava';
+import avaRuleTester from 'eslint-ava-rule-tester';
+import {outdent} from 'outdent';
+import rule from '../rules/prefer-set-has';
+
+const ruleId = 'prefer-set-has';
+
+const ruleTester = avaRuleTester(test, {
+	parserOptions: {
+		ecmaVersion: 2020,
+		sourceType: 'module'
+	}
+});
+
+const createError = name => [
+	{
+		messageId: 'preferSetHas',
+		data: {
+			name
+		}
+	}
+];
+
+ruleTester.run(ruleId, rule, {
+	valid: [
+		outdent`
+			const foo = new Set([1, 2, 3]);
+			const exists = foo.has(1);
+		`,
+		// Not `VariableDeclarator`
+		outdent`
+			foo = [1, 2, 3];
+			const exists = foo.includes(1);
+		`,
+		outdent`
+			const exists = foo.includes(1);
+		`,
+		outdent`
+			const exists = [1, 2, 3].includes(1);
+		`,
+		// Didn't call `includes()`
+		outdent`
+			const foo = [1, 2, 3];
+		`,
+		// Not `CallExpression`
+		outdent`
+			const foo = [1, 2, 3];
+			const exists = foo.includes;
+		`,
+		// Not `foo.includes()`
+		outdent`
+			const foo = [1, 2, 3];
+			const exists = includes(foo);
+		`,
+		outdent`
+			const foo = [1, 2, 3];
+			const exists = bar.includes(foo);
+		`,
+		outdent`
+			const foo = [1, 2, 3];
+			const exists = foo[includes](1);
+		`,
+		outdent`
+			const foo = [1, 2, 3];
+			const exists = foo.indexOf(1) !== -1;
+		`,
+		// Not only `foo.includes()`
+		outdent`
+			const foo = [1, 2, 3];
+			const exists = foo.includes(1);
+			foo.length = 1;
+		`,
+		outdent`
+			const foo = [1, 2, 3];
+			function unicorn() {
+				if (foo.includes(1)) {}
+				return foo;
+			}
+		`,
+		// Declared more than once
+		outdent`
+			var foo = [1, 2, 3];
+			var foo = [4, 5, 6];
+			const exists = foo.includes(1);
+		`,
+
+		// Not `ArrayExpression`, maybe we can enable some of those later
+		outdent`
+			const foo = bar;
+			const exists = foo.includes(1);
+		`,
+		outdent`
+			const foo = [1, 2, 3].slice();
+			const exists = foo.includes(1);
+		`,
+		outdent`
+			const foo = Array.from(bar);
+			const exists = foo.includes(1);
+		`,
+
+		// Extra arguments
+		outdent`
+			const foo = [1, 2, 3];
+			const exists = foo.includes();
+		`,
+		outdent`
+			const foo = [1, 2, 3];
+			const exists = foo.includes(1, 1);
+		`,
+		outdent`
+			const foo = [1, 2, 3];
+			const exists = foo.includes(1, 0);
+		`,
+		outdent`
+			const foo = [1, 2, 3];
+			const exists = foo.includes(1, undefined);
+		`,
+		outdent`
+			const foo = [1, 2, 3];
+			const exists = foo.includes(...[1]);
+		`,
+		// TODO: enable this test when eslint support optional-chaining
+		// Optional
+		// outdent`
+		// 	const foo = [1, 2, 3];
+		// 	const exists = foo.?includes(1);
+		// `,
+		// Different scope
+		outdent`
+			function unicorn() {
+				const foo = [1, 2, 3];
+			}
+			const exists = foo.includes(1);
+		`,
+
+		outdent`
+			export const foo = [1, 2, 3];
+			const exists = foo.includes(1);
```

Comment:
Is there any other way can expose a variable need  to be  handled?

Reply:
Sorry, I don't understand the question.

Reply:
We need check variable only called `.includes` and make sure it can't modify from other file. I think `export const foo` is the only way to make foo exposed without accessing `foo`?

Reply:
Maybe this?

```js
module.exports = [1, 2, 3];
const exists = module.exports.includes(1);
```

---

Can you add tests for these cases too? Just to ensure they're covered.

```js
const foo = [1, 2, 3];
export {foo};
const exists = foo.includes(1);
```

```js
const foo = [1, 2, 3];
export default foo;
const exists = foo.includes(1);
```

```js
const foo = [1, 2, 3];
export {foo as bar};
const exists = foo.includes(1);
```

```js
const foo = [1, 2, 3];
module.exports = foo;
const exists = foo.includes(1);
```

```js
const foo = [1, 2, 3];
exports = foo;
const exists = foo.includes(1);
```

```js
const foo = [1, 2, 3];
module.exports.foo = foo;
const exists = foo.includes(1);
```

Reply:
```js
module.exports = [1, 2, 3];
const exists = module.exports.includes(1);
```

This won't reported, because only `VariableDeclaration` is allowed. Other tests added https://github.com/sindresorhus/eslint-plugin-unicorn/pull/604/commits/3282cccdc83f6bbae6fb5e2a192312f591333e5b .